yt/geometry/particle_geometry_handler.py

```diff
@@ -159,43 +207,57 @@ def _initialize_index(self):
         # Load Morton index from file if provided
         fname = getattr(ds, "index_filename", None) or f"{ds.parameter_filename}.ewah"
 
+        self.pii = ParticleIndexInfo(order1, order2, fname, mutable_index)
+
         self.regions = ParticleBitmap(
             ds.domain_left_edge,
             ds.domain_right_edge,
             ds.periodicity,
             self.ds._file_hash,
             len(self.data_files),
-            index_order1=order1,
-            index_order2=order2,
+            index_order1=self.pii.order1,
+            index_order2=self.pii.order2_orig,
         )
 
         dont_load = dont_cache and not hasattr(ds, "index_filename")
         try:
             if dont_load:
                 raise OSError
-            rflag = self.regions.load_bitmasks(fname)
+            rflag, max_hsml = self.regions.load_bitmasks(fname)
+            if max_hsml > 0.0 and self.pii.mutable_index:
+                self._order2_update(max_hsml)
             rflag = self.regions.check_bitmasks()
             self._initialize_frontend_specific()
             if rflag == 0:
                 raise OSError
         except (OSError, struct.error):
             self.regions.reset_bitmasks()
-            self._initialize_coarse_index()
+            max_hsml = self._initialize_coarse_index()
             self._initialize_refined_index()
-            # We now update fname since index_order2 may have changed
-            fname = (
-                getattr(ds, "index_filename", None) or f"{ds.parameter_filename}.ewah"
-            )
             wdir = os.path.dirname(fname)
             if not dont_cache and os.access(wdir, os.W_OK):
                 # Sometimes os mis-reports whether a directory is writable,
                 # So pass if writing the bitmask file fails.
                 try:
-                    self.regions.save_bitmasks(fname)
+                    self.regions.save_bitmasks(fname, max_hsml)
                 except OSError:
                     pass
             rflag = self.regions.check_bitmasks()
 
+        self.ds.index_order = (self.pii.order1, self.pii.order2)
```

Comment:
I think I agree with all of this -- but here's where I start to wonder if having a separate class instead of properties would be better.  I wonder if proxying `index_order` as a property that defers to `self.pii` might reduce the level of indirect?

Reply:
I had a look at this. The issue is that we pass this in to `yt.load` and immediately make it a property of `Dataset` (or more usually `ParticleDataset`). That's how it's accessed in the first place when we construct the `ParticleIndex` instance. So instead we'd have to pass it through `Dataset` to the index construction. I'm fine with this in principle, but it does mean adding a bunch of arguments to other functions and methods so we can pass the value down.

Reply:
Nah, let's skip it then.